src/Analyzer/Passes/UniqToCountPass.cpp

```diff
@@ -0,0 +1,142 @@
+{
+
+namespace
+{
+
+bool matchFnUniq(String func_name)
+{
+    auto name = Poco::toLower(func_name);
+    return name == "uniq" || name == "uniqHLL12" || name == "uniqExact" || name == "uniqTheta" || name == "uniqCombined"
+        || name == "uniqCombined64";
+}
+
+bool nodeEquals(const QueryTreeNodePtr & lhs, const QueryTreeNodePtr & rhs)
+{
+    auto * lhs_node = lhs->as<ColumnNode>();
+    auto * rhs_node = rhs->as<ColumnNode>();
+
+    if (lhs_node && rhs_node && lhs_node->getColumn() == rhs_node->getColumn())
+        return true;
+    return false;
+}
+
+bool nodeListEquals(const QueryTreeNodes & lhs, const QueryTreeNodes & rhs)
+{
+    if (lhs.size() != rhs.size())
+        return false;
+    for (size_t i = 0; i < lhs.size(); i++)
+    {
+        if (!nodeEquals(lhs[i], rhs[i]))
+            return false;
+    }
+    return true;
+}
+
+bool nodeListContainsAll(const QueryTreeNodes & lhs, const QueryTreeNodes & rhs)
+{
+    if (lhs.size() < rhs.size())
+        return false;
+    for (const auto & re : rhs)
+    {
+        auto predicate = [&](const QueryTreeNodePtr & le) { return nodeEquals(le, re); };
+        if (std::find_if(lhs.begin(), lhs.end(), predicate) == lhs.end())
+            return false;
+    }
+    return true;
+}
+
+}
+
+class UniqToCountVisitor : public InDepthQueryTreeVisitorWithContext<UniqToCountVisitor>
+{
+public:
+    using Base = InDepthQueryTreeVisitorWithContext<UniqToCountVisitor>;
+    using Base::Base;
+
+    void visitImpl(QueryTreeNodePtr & node)
+    {
+        if (!getSettings().optimize_uniq_to_count)
+            return;
+
+        auto * query_node = node->as<QueryNode>();
+        if (!query_node)
+            return;
+
+        /// Check that query has only single table expression which is subquery
+        auto * subquery_node = query_node->getJoinTree()->as<QueryNode>();
+        if (!subquery_node)
+            return;
+
+        /// Check that query has only single node in projection
+        auto & projection_nodes = query_node->getProjection().getNodes();
+        if (projection_nodes.size() != 1)
+            return;
+
+        /// Check that projection_node is a function
+        auto & projection_node = projection_nodes[0];
+        auto * function_node = projection_node->as<FunctionNode>();
+        if (!function_node)
+            return;
+
+        /// Check that query single projection node is `uniq` or its variants
+        if (!matchFnUniq(function_node->getFunctionName()))
+            return;
+
+        auto & uniq_arguments_nodes = function_node->getArguments().getNodes();
+
+        /// Whether query matches 'SELECT uniq(x ...) FROM (SELECT DISTINCT x ...)'
+        auto match_subquery_with_distinct = [&]() -> bool
+        {
+            if (!subquery_node->isDistinct())
+                return false;
+            /// uniq expression list == subquery group by expression list
+            if (!nodeListEquals(uniq_arguments_nodes, subquery_node->getProjection().getNodes()))
```

Comment:
When an alias is set in the projection column of a subquery and this alias is used in the uniq function of the outer query, the columns of the subquery and the outer query will not be matched, which will lead to rewriting fail. For example the following SQL:
```
explain query tree SELECT uniq(alias_of_a) FROM (SELECT a as alias_of_a, sum(b) FROM test_rewrite_uniq_to_count GROUP BY a) t settings optimize_uniq_to_count = 1, allow_experimental_analyzer = 1;
```

The reason is that the alias in the outer query `uniq(alias_of_a)` is not expanded. I think we should expand all alias in query after `QueryAnalysis`.